Add liquidity checks and improve payment error handling

slanesuke opened the pull request:
This PR introduces a few improvements including liquidity checks in the `pay_for_offer` and `create_refund_builder`,  updated error handling, and support for fiat-denominated offers.
#### Changes:
* Liquidity Checks:
    * Added early aborts in `pay_for_offer` and `create_refund_builder` if channel liquidity is insufficient.
    * Added tests to validate the new liquidity checks.
* Error Handling:
    * Introduced `Bolt12RequestError` for errors related to BOLT12 payment/refund requests.
    * Introduced `Bolt12CreationError` for errors occurring during the creation of BOLT12 offers/refunds.
* Fiat-Denominated Offers:
    * Improved support for offers specifying an amount in fiat currencies.
    * Added tests for `InvoiceRequest` creation with currency-based amounts.
* Docs updates:
    * Updated the docs for `InvoiceRequestBuilder::amount_msats` to clarify behavior when handling currency-based offers.
    * Updated the docs for `Bolt12SemanticError::InvalidAmount` to include cases where the invoice amount does not match the expected amount.

Fixes #3174

slanesuke committed:
Abort payment if insufficient liquidity in `pay_for_offer`

Check for sufficient channel liquidity in `pay_for_offer` and
abort the payment if there is insufficient liquidity to fulfill
the payment. This makes sure that the payment attempt is only
made when there is enough outbound liquidity available, preventing
failed payment attempts due to liquidity issues.

slanesuke commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8816,6 +8816,26 @@ where
+				log_error!(self.logger, "Insufficient liquidity for payment with payment id: {}", payment_id);
+				Err(Bolt12SemanticError::InsufficientLiquidity)
+			} else {
+				log_error!(self.logger, "Offer amount is missing for payment with id: {}", payment_id);
```
Much simpler. Thanks for the feedback!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8816,8 +8816,15 @@ where
 		let invoice_request = builder.build_and_sign()?;
 
+		let total_liquidity: u64 = self.list_usable_channels().iter().map(|channel| channel.next_outbound_htlc_limit_msat).sum();
+		if invoice_request.amount_msats() > Some(total_liquidity) {
```
This will fail in the case where we have an offer with a value denominated in bitcoin and caller doesn't provide an amount. We should use the value in the offer in that case.

slanesuke replied:
Got it updated!

tnull reviewed (dismissed):
LGTM, but will let @jkczyz take another look.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8818,6 +8818,22 @@ where
+			},
+		};
+
+		if let Some(amount) = total_amount_msats {
+			if amount > total_liquidity {
+				log_error!(self.logger, "Insufficient liquidity for payment with payment id: {}", payment_id);
+				return Err(Bolt12SemanticError::InsufficientLiquidity);
+			}
+		}
```
Let's add a test to `offers_tests.rs` to check this error. Looks like one of those tests is failing when hitting this case, so we'll likely need to modify that test setup.

jkczyz replied:
Hmmm... the test is checking `DefaultMessageRouter`'s  behavior, so it would be nice to keep that coverage. I think the only way to bypass the liquidity check is give the offer a fiat denomination using `OfferBuilder::amount`  (making it `pub(crate)`) instead of `OfferBuilder::amount_msats`.

jkczyz replied:
So I was helping @slanesuke debug why this isn't working, and it seems we currently don't support _paying_ fiat-denominated offers, which is unexpected -- we should only not support _creating_ them. `InvoiceRequestBuilder::build_and_sign` is failing because it eventually calls `OfferContents::check_amount_msats_for_quantity`, which is returning `UnsupportedCurrency`.

That function is called when parsing an `InvoiceRequest`, too, so we'd want to fail there, I think. Maybe we move the currency check to the parsing code and let the builder code trust the msats amount set by the user?

jkczyz replied:
> That function is called when parsing an `InvoiceRequest`, too, so we'd want to fail there, I think. Maybe we move the currency check to the parsing code and let the builder code trust the msats amount set by the user?

@slanesuke Let me know if you can take this on. The solution above should be right.

jkczyz commented on lightning/src/offers/parse.rs:
```diff
@@ -193,6 +193,8 @@ pub enum Bolt12SemanticError {
 	UnexpectedPaymentHash,
 	/// A signature was expected but was missing.
 	MissingSignature,
+	/// There is insufficient liquidity to complete the payment.
+	InsufficientLiquidity,
```
I'm a little hesitant adding another variant as it's not really a semantic error. Admittedly, we've sort of stretched the use of `Bolt12SemanticError` to be used with the builder functions (e.g., using `MissingPaths` when we can't *create* a reply path) instead of creating an error type that wraps this to support other variants. Do other reviewer have any opinions on this? (@tnull @TheBlueMatt)

jkczyz replied:
We already have `Bolt12ParseError` which has a variant that wraps `Bolt12SemanticError`.

slanesuke replied:
Is there an existing alternative error type that would be better for this case?

jkczyz replied:
The only thing close would be `InsufficientAmount`, but using it would also be stretching its intended use.

jkczyz replied:
Discussed offline with @tnull. Let's use a new error type in the `ChannelManager` functions (maybe `Bolt12CreationError`?) that is an enum with one variant wrapping `Bolt12SemanticError` and other variants for errors that can occur outside of the builder (e.g., failing to create a blinded path, not enough liquidity, duplicate payment id, etc.). See `Bolt12ParseError` for an example. This may require removing variants like `DuplicatePaymentId` from `Bolt12SemanticError`.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -64,7 +64,7 @@ use crate::ln::wire::Encode;
 use crate::offers::invoice_error::InvoiceError;
 use crate::offers::invoice_request::{DerivedPayerId, InvoiceRequestBuilder};
-use crate::offers::offer::{Offer, OfferBuilder};
+use crate::offers::offer::{Offer, OfferBuilder, Amount};
```
Use lexicographic order.

jkczyz reviewed (commented):
Drop the fixup message from the commit message.

slanesuke committed:
Refactor amount validation and relocate currency check

Update `InvoiceRequest` parsing to handle currency amounts in
offers. The currency check has been moved from
`OfferContents::check_amount_msats_for_quantity` to the
`InvoiceRequest` parsing code. This makes sure that an error is
returned early if an offer specifies an unsupported currency,
preventing issues with parsing a `Bolt12Invoice`.

This change also simplifies
`OfferContents::check_amount_msats_for_quantity` by removing
redundant currency validation, allowing it to trust the
sender-provided amount.

slanesuke commented:
@jkczyz I added a test to check the `Insufficient Liquidity` error in `offers_tests`.   Moved the currency check to the parsing code and modified the amount check to trust the user-provided msats amount (I think I did this correctly). 

And regarding the InsufficientLiquidity error type, I left it as is. I just saw the new comments so I'll add the `Bolt12CreationError` now!

slanesuke commented:
> @jkczyz I added a test to check the `Insufficient Liquidity` error in `offers_tests`. Moved the currency check to the parsing code and modified the amount check to trust the user-provided msats amount (I think I did this correctly).
> 
> And regarding the InsufficientLiquidity error type, I left it as is. I just saw the new comments so I'll add the `Bolt12CreationError` now!

Hmmm... looks like a few different test are failing now.

slanesuke committed:
Add `Bolt12CreationError` error type

Introduce the `Bolt12CreationError` error type for handling
BOLT12-related errors in the `ChannelManager`. This error
type consolidates various variants, which were previously
part of `Bolt12SemanticError`. Additionally, updated the
relevant code to replace `Bolt12SemanticError` with
`Bolt12CreationError` throughout the affected files.

Note: The following commit will revert the changes in the
`create_refund_builder` and `pay_for_offer` methods, switching
back to `Bolt12SemanticError` temporarily. A future commit
will introduce a `Bolt12RequestError` to handle these scenarios
more accurately.

slanesuke commented:
The variant wrapping `Bolt12SemanticError`  could use a better name and the docs for `Bolt12CreationError` may need some fine-tuning. But this should be close to finished.

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -1131,13 +1131,15 @@ impl TryFrom<PartialInvoiceRequestTlvStream> for InvoiceRequestContents {
+		let amount_msats = match amount {
+			Some(amount) => Some(amount),
+			None => match offer.amount() {
+				Some(Amount::Bitcoin { amount_msats }) => Some(amount_msats),
```
This would cause us to set the amount when the parsed `InvoiceRequest` didn't have one. I think we only need to call `check_amount_msats_for_quantity` when we have `Some(amount)`.

slanesuke replied:
I noticed that calling `check_amount_msats_for_quantity` only when we have `Some(amount)` causes the test `parses_invoice_request_with_amount` to fail. It seems we need to call it even when amount is `None`. Skipping the `None` check misses the validation for the offer's currency, so the test doesn't catch the `UnsupportedCurrency` error.

So for now, I kept the call to `check_amount_msats_for_quantity` regardless of whether amount is Some or None. Let me know if this approach is correct!

jkczyz replied:
Right, we should still call that, but as written my earlier point still stands. We are now setting `Some(amount_msats)` on the `InvoiceRequestContents` below when we shouldn't.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2477,6 +2477,26 @@ pub enum RecentPaymentDetails {
+#[derive(Debug, Clone, PartialEq)]
+pub enum Bolt12CreationError {
+    /// Error from  BOLT 12 semantic checks.
+    Bolt12SendingError(Bolt12SemanticError),
```
`InvalidSemantics`

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2477,6 +2477,26 @@ pub enum RecentPaymentDetails {
+    /// There is insufficient liquidity to complete the payment.
+    InsufficientLiquidity,
+    /// Failed to create a blinded path.
+    FailedToCreateBlindedPath
```
`BlindedPathCreationFailed`

jkczyz replied:
End with comma.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8842,7 +8862,7 @@ where
 		payer_note: Option<String>, payment_id: PaymentId, retry_strategy: Retry,
 		max_total_routing_fee_msat: Option<u64>
-	) -> Result<(), Bolt12SemanticError> {
+	) -> Result<(), Bolt12CreationError> {
```
Could you update the other BOLT12-related `ChannelManager` methods to use this error, too? Ideally, change one per commit.

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -1131,7 +1131,17 @@ impl TryFrom<PartialInvoiceRequestTlvStream> for InvoiceRequestContents {
+			None => match offer.amount() {
+				Some(Amount::Bitcoin { amount_msats }) => Some(amount_msats),
+				Some(Amount::Currency { .. }) => {
+					return Err(Bolt12SemanticError::UnsupportedCurrency)
```
End with semi-colon.

slanesuke commented:
@jkczyz ready for review.

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -1267,15 +1268,15 @@ fn fails_creating_or_paying_for_offer_without_connected_peers() {
 	match david.node.pay_for_offer(&offer, None, None, None, payment_id, Retry::Attempts(0), None) {
 		Ok(_) => panic!("Expected error"),
-		Err(e) => assert_eq!(e, Bolt12SemanticError::InsufficientLiquidity),
+		Err(e) => assert_eq!(e, Bolt12SemanticError::MissingPaths),
```
Let's re-order the commits to avoid changing this back and forth.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2481,18 +2481,18 @@ pub enum RecentPaymentDetails {
 pub enum Bolt12CreationError {
     /// Error from  BOLT 12 semantic checks.
-    Bolt12SendingError(Bolt12SemanticError),
+    InvalidSemantics(Bolt12SemanticError),
```
Commits should be atomic. Let's make a fixup commit for the previous commit (to squash later) for all the renames. `git add -p` is your friend. 🙂

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -1130,18 +1130,13 @@ impl TryFrom<PartialInvoiceRequestTlvStream> for InvoiceRequestContents {
-				None => None,
-			}
+		match offer.amount() {
+			None => 0,
+			Some(Amount::Bitcoin {amount_msats}) => amount_msats,
```
These looks like a no-op?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8672,15 +8672,15 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
 		let path = $self.create_blinded_paths_using_absolute_expiry(OffersContext::Unknown {}, absolute_expiry)
 			.and_then(|paths| paths.into_iter().next().ok_or(()))
-			.map_err(|_| Bolt12SemanticError::MissingPaths)?;
+			.map_err(|_| Bolt12CreationError::InvalidSemantics(Bolt12SemanticError::MissingPaths))?;
```
Should be `FailedToCreateBlindedPath`. Likewise in other commits.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1654,10 +1654,9 @@ where
+/// # use lightning::ln::channelmanager::{AChannelManager, Bolt12CreationError};
 /// #
-/// # fn example<T: AChannelManager>(channel_manager: T) -> Result<(), Bolt12SemanticError> {
+/// # fn example<T: AChannelManager>(channel_manager: T) -> Result<(), Bolt12CreationError> {
```
Please do this in a fixup commit. We require tests to pass on each commit after all fixups are squashed.

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -1260,7 +1260,7 @@ fn fails_creating_or_paying_for_offer_without_connected_peers() {
 	match alice.node.create_offer_builder(Some(absolute_expiry)) {
 		Ok(_) => panic!("Expected error"),
-		Err(e) => assert_eq!(e, Bolt12SemanticError::MissingPaths),
+		Err(e) => assert_eq!(e, Bolt12CreationError::InvalidSemantics(Bolt12SemanticError::MissingPaths)),
```
Likewise, these changes should go in the commits that added `Bolt12CreationError`.

slanesuke committed:
Add test for insufficient liquidity in `pay_for_offer`

Adds the `fails_paying_offer_with_insufficient_liquidity` test to
ensure that the `pay_for_offer` method correctly handles cases
where the liquidity is insufficient. The test verifies that the
method returns the `Bolt12RequestError::InsufficientLiquidity`
error when the requested payment amount exceeds the available
liquidity.

jkczyz commented on lightning/src/offers/offer.rs:
```diff
@@ -857,9 +857,8 @@ impl OfferContents {
-			Some(Amount::Bitcoin { amount_msats }) => amount_msats,
-			Some(Amount::Currency { .. }) => return Err(Bolt12SemanticError::UnsupportedCurrency),
+			Some(Amount::Bitcoin {amount_msats}) => amount_msats,
+			_ => 0,
```
Seems we should have a unit test that given an `Offer` with a currency, we (1) successfully create an `InvoiceRequest` when not providing an amount and (2) fail creating an `InvoiceRequest` when providing an amount. We can probably update / extend `builds_invoice_request_with_amount` as needed.

jkczyz replied:
Ok, I re-checked the spec. It's fine to set the amount in an `InvoiceRequest` when the `Offer` issuer specifies a currency. The issuer will reject it if they disagree and otherwise MUST set a matching amount in the invoice.

jkczyz replied:
Hmm... wow, looks like we don't. 😕  Let's fix that in this PR or a follow-up.

slanesuke replied:
Sounds good, I'll just make the changes in this PR.

slanesuke committed:
Add tests for `Offers` with fiat amount

Introduces tests to ensure the successful creation of
`InvoiceRequest` when specifying a currency amount with
no amount_msats. And verifies that when both a currency
amount and amount_msats are specified, the values are handled
correctly. This ensures that `InvoiceRequest` behaves as
expected with various amount configs.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8991,7 +9009,7 @@ where
 		payer_note: Option<String>, payment_id: PaymentId, retry_strategy: Retry,
 		max_total_routing_fee_msat: Option<u64>
-	) -> Result<(), Bolt12SemanticError> {
+	) -> Result<(), Bolt12CreationError> {
```
Right... could use a better name. Was about to suggest `Bolt12PaymentError`, but we already have that when paying the invoice. Maybe `Bolt12RequestError`?

jkczyz replied:
Ah, yeah we can call it `Bolt12CreationError` for `create_offer_builder` then.  Presumably, we should check liquidity when calling `create_refund_builder`, too. That along with `pay_for_offer` can use `Bolt12RequestError`, or something more appropriately named.

slanesuke replied:
Sounds good.

slanesuke commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9038,6 +9038,22 @@ where
 		let reply_paths = self.create_blinded_paths(context)
 			.map_err(|_| Bolt12CreationError::BlindedPathCreationFailed)?;
 
+		let total_liquidity: u64 = self.list_usable_channels().iter().map(|channel| channel.next_outbound_htlc_limit_msat).sum();
```
Gotcha, thanks for the tip! I'll make the change.

slanesuke replied:
Ahhh shoot! Okay, thanks for your patience :sweat_smile:

slanesuke replied:
> `list_channels` is even worse :). Instead iterate over the `Channel`s themselves and call `get_available_balances` on them (see `ChanelDetails::from_channel_context`).

I having trouble figuring out how to iterate directly over the list of channels rather than using the `ChannelDetails` list! Could you point me in the right direction on how to access and iterate over the channels themselves?

slanesuke committed:
Abort `RefundBuilder` creation on insufficient liquidity

Add a liquidity check in `create_refund_builder` to ensure
that refund creation is aborted if there is insufficient
outbound liquidity. This change prevents attempts to create
a refund when the available liquidity is below the required
amount, ensuring that refunds are only created when there
is adequate liquidity.

slanesuke committed:
Add a `Bolt12RequestError` error type

Introduced the `Bolt12RequestError` type to handle errors
associated with BOLT12 payment and refund requests. This
type includes variants for handling `InvalidSemantics`,
`DuplicatePaymentId`, `InsufficientLiquidity`, and
`BlindedPathCreationFailed`.

slanesuke committed:
Validate amount_msats against invoice and refund amounts

Add a check to ensure that the amount_msats in an invoice
matches the amount_msats specified in the invoice_request
or refund. Reject the invoice as invalid if there is a
mismatch between these amounts. This validation ensures
consistency in invoice handling.

slanesuke committed:
Update docs for `InvoiceRequestBuilder::amount_msats`

Improves documentation for the amount_msats method in 
`InvoiceRequestBuilder`. The update describes the
behavior when an `Offer` specifies a currency.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9050,7 +9112,37 @@ where
 		let context = OffersContext::OutboundPayment { payment_id, nonce, hmac: Some(hmac) };
 		let reply_paths = self.create_blinded_paths(context)
-			.map_err(|_| Bolt12SemanticError::MissingPaths)?;
+			.map_err(|_| Bolt12CreationError::BlindedPathCreationFailed)?;
```
We can use `Bolt12RequestError` here, that way we don't need the `From<Bolt12CreationError>` implementation.

slanesuke replied:
Oops, missed that one. Thanks!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9050,7 +9112,37 @@ where
+			sum
+		};
+
+		let total_amount_msats = match invoice_request.amount_msats() {
```
How about `requested_amount_msats`?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9050,7 +9112,37 @@ where
+				let peer_state_lock = peer_state_mutex.lock().unwrap();
+				let peer_state = &*peer_state_lock;
+
+				peer_state.channel_by_id.iter().for_each(|(_, phase)| {
+					sum += phase.context().get_available_balances(&self.fee_estimator).outbound_capacity_msat;
+				});
```
nit: Instead of `for_each()`, you could use `map().sum()` and add the result to `sum`.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9059,7 +9151,7 @@ where
 				payment_id, expiration, retry_strategy, max_total_routing_fee_msat
 			)
-			.map_err(|_| Bolt12SemanticError::DuplicatePaymentId)?;
+			.map_err(|_| Bolt12RequestError::DuplicatePaymentId)?;
```
Let's use `()` instead of `_` here and in other places where `.map_err` is used. That way, if the error type ever is changed we are forced to reconsider what the proper mapping should be.

slanesuke replied:
Okay, good to know!

jkczyz commented on lightning/src/offers/invoice.rs:
```diff
@@ -1344,6 +1344,13 @@ impl TryFrom<PartialInvoiceTlvStream> for InvoiceContents {
 				(payer_tlv_stream, offer_tlv_stream, invoice_request_tlv_stream)
 			)?;
+
+			if let Some(requested_amount_msats) = invoice_request.amount_msats() {
+				if amount_msats != requested_amount_msats {
+					return Err(Bolt12SemanticError::UnexpectedAmount);
+				}
+			}
```
Let's also do this for the refund case above.

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -270,6 +270,11 @@ macro_rules! invoice_request_builder_methods { (
 	/// Sets the [`InvoiceRequest::amount_msats`] for paying an invoice. Errors if `amount_msats` is
 	/// not at least the expected invoice amount (i.e., [`Offer::amount`] times [`quantity`]).
 	///
+	/// When the [`Offer`] specifies a currency, this method allows setting the `amount_msats` in the
+	/// `InvoiceRequest`. The issuer of the offer will validate this amount and must set a matching
+	/// amount in the final invoice. If the issuer disagrees with the specfified amount, the invoice
+	/// request will be rejected.
```
Also useful if the `Offer` doesn't have any amount.

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -1725,6 +1735,26 @@ mod tests {
 		}
+
+		let invoice_request = OfferBuilder::new(recipient_pubkey())
+			.amount(Amount::Currency {iso4217_code: *b"USD", amount: 1000})
```
Spaces between curly braces and identifiers / expressions. Same elsewhere.

jkczyz commented:
> I’ve made the changes to add both `Bolt12CreationError` and `Bolt12RequestError` as separate error types. Since both share a couple common variants (`InvalidSemantics` and `BlindedPathCreationFailed`), I’m wondering about redundancy. Would you guys recommend keeping the current approach with these duplicated variants, or do you think there’s a better way to handle this to avoid redundancy?

Nah, the redundancy is fine here, IMO.

> Also, I’ve updated the approach to calculate the total channel liquidity by (hopefully) iterating directly over the channels themselves. Would this be correct/efficient, or are there better ways to accomplish this!?
> 
> (cc @jkczyz @TheBlueMatt)

Yeah, I think this is what @TheBlueMatt had in mind.

slanesuke commented:
> > I’ve made the changes to add both `Bolt12CreationError` and `Bolt12RequestError` as separate error types. Since both share a couple common variants (`InvalidSemantics` and `BlindedPathCreationFailed`), I’m wondering about redundancy. Would you guys recommend keeping the current approach with these duplicated variants, or do you think there’s a better way to handle this to avoid redundancy?
> 
> Nah, the redundancy is fine here, IMO.
> 
> > Also, I’ve updated the approach to calculate the total channel liquidity by (hopefully) iterating directly over the channels themselves. Would this be correct/efficient, or are there better ways to accomplish this!?
> > (cc @jkczyz @TheBlueMatt)
> 
> Yeah, I think this is what @TheBlueMatt had in mind.

Sweet, thanks for the feedback! Just pushed the updates.

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -1159,8 +1159,13 @@ impl TryFrom<PartialInvoiceRequestTlvStream> for InvoiceRequestContents {
-		offer.check_quantity(quantity)?;
-		offer.check_amount_msats_for_quantity(amount, quantity)?;
+		match offer.amount() {
+			Some(Amount::Currency {..}) => return Err(Bolt12SemanticError::UnsupportedCurrency),
```
Unfortunately, it looks like we can't fail like this because this `TryFrom` implementation is also used when parsing a `Bolt12Invoice`. So if we send an amount-less request for an `Offer` with a currency, we'll fail to parse the invoice that is sent back. Here's a [branch](https://github.com/jkczyz/rust-lightning/commits/2024-08-invoice-from-offer-with-currency/) with a test that demonstrates this. Feel free to cherrypick the commit.

I guess we need to move the these checks higher up in the parsing code, after the contents has been created. Haven't checked if that will have any unexpected behavior.

jkczyz replied:
@TheBlueMatt @tnull I wonder if we should allow parsing an `InvoiceRequest` with any amount and rather check the amount when handling it (i.e., in `ChannelManager`). Otherwise, once currency support is added, it would require looking up exchange rates at parsing time, which seems wrong. Also, if no amount is specified in the request, then that is where the issuer would convert from the offer amount. Thoughts?

slanesuke replied:
@jkczyz I moved the checks higher up in the parsing code as you suggested. But, after cherry-picking the commit and running the tests, the `parses_invoice_with_currency` test fails. Given the context, it looks like checking amounts during handling might be more fitting. If so, I can include this change in this PR.

jkczyz replied:
Sure, let's go ahead and do that. Longer term it makes more sense, IMO.

jkczyz commented on lightning/src/offers/invoice.rs:
```diff
@@ -1339,11 +1339,23 @@ impl TryFrom<PartialInvoiceTlvStream> for InvoiceContents {
 			)?;
+
+			if amount_msats != refund.amount_msats() {
+				return Err(Bolt12SemanticError::UnexpectedAmount);
+			}
+
 			Ok(InvoiceContents::ForRefund { refund, fields })
 		} else {
 			let invoice_request = InvoiceRequestContents::try_from(
 				(payer_tlv_stream, offer_tlv_stream, invoice_request_tlv_stream)
 			)?;
+
+			if let Some(requested_amount_msats) = invoice_request.amount_msats() {
+				if amount_msats != requested_amount_msats {
+					return Err(Bolt12SemanticError::UnexpectedAmount);
```
These should be `InvalidAmount` since `UnexpectedAmount` indicates an amount was provided but was not expected to be set. Should also updated the docs on `InvalidAmount` to reflect this usage.

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -1730,6 +1730,26 @@ mod tests {
+			.request_invoice(vec![1; 32], payer_pubkey()).unwrap()
+			.build().unwrap()
+			.sign(payer_sign).unwrap();
+		assert_eq!(invoice_request.amount(), Some(Amount::Currency { iso4217_code: *b"USD", amount: 1000 }));
```
Shouldn't we assert that amount is `None` on `invoice_request` and the invreq TLV stream?

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -1730,6 +1730,26 @@ mod tests {
 			Err(e) => assert_eq!(e, Bolt12SemanticError::InvalidAmount),
 		}
+
+		let invoice_request = OfferBuilder::new(recipient_pubkey())
```
What is meant by "However, the return behavior isn't what's expected; its a work in progress." in commit 73f643326dbf187a69423c5687c8e7b2542e6503?

slanesuke replied:
Ah, at that point, I was expecting the unit test to fail when creating an `InvoiceRequest` with an `Offer` that includes a currency and an amount, but I wasn’t sure why it didn’t. I should've updated the commit message after we clarified the behavior and I changed the test to succeed instead of fail.

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -2182,3 +2182,34 @@ fn fails_paying_invoice_with_unknown_required_features() {
 }
+
+#[test]
+fn fails_paying_offer_with_insufficient_liquidity() {
```
Could you add one for creating refunds, too?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8905,7 +8923,7 @@ macro_rules! create_refund_builder { ($self: ident, $builder: ty) => {
 		&$self, amount_msats: u64, absolute_expiry: Duration, payment_id: PaymentId,
 		retry_strategy: Retry, max_total_routing_fee_msat: Option<u64>
-	) -> Result<$builder, Bolt12SemanticError> {
+	) -> Result<$builder, Bolt12CreationError> {
```
Let's not change this to `Bolt12CreationError` if a few commits later it is changed to `Bolt12RequestError`. You can insert a fixup commit here to revert the change, which can be squashed before merging. Similarly for `pay_for_offer`.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9050,7 +9104,38 @@ where
+			},
+		};
+
+		if let Some(amount) = requested_amount_msats {
+			if amount > total_liquidity {
+				log_error!(self.logger, "Insufficient liquidity for payment with payment id: {}", payment_id);
+				return Err(Bolt12RequestError::InsufficientLiquidity);
+			}
+		}
```
@tnull Should we consider outstanding invoice requests / refunds?

jkczyz replied:
Not sure I understand the point. For BOLT11, if we initiate a bunch of payments, each subsequent one will have less liquidity to consider, IIUC. But for BOLT12, we can similarly initiate more than one payment but each subsequent one will consider the same liquidity unless an invoice is received for a prior one.

slanesuke replied:
Sorry for the late reply! Since outstanding invoice requests and refunds can tie up liquidity, do you think we should check for any outstanding invoices and see how they affect available liquidity before moving forward with the payment? And should I address it in this PR? I want to make sure to cover all our bases.

jkczyz replied:
So they can potentially _use_ liquidity if an invoice is received in response. But they don't _tie up_ liquidity since a payment hasn't been made yet. That said, we can consider a request for payment as an intention to tie up liquidity in the hopefully near future. I'd lean towards including it but will defer to @tnull.

FWIW, there's also some edge cases to consider. If we have a fiat-denominated offer and the users *doesn't* give an amount, how much liquidity should we considered unavailable?

jkczyz reviewed (commented):
Regarding commit messages, please state not only the _what_ but also the _why_ (but not the _how_). Also, use the imperative instead mood rather than first person (e.g., "Moved" vs "I moved"). See the following guidelines: https://cbea.ms/git-commit/.

slanesuke committed:
f Revert `Bolt12CreationError` to `Bolt12SemanticError`

Revert the return type of `create_refund_builder` and
`pay_for_offer` from `Bolt12CreationError` back to
`Bolt12SemanticError`. This change is temporary, as a
future commit will introduce `Bolt12RequestError` for
these functions.

slanesuke committed:
Update docs for `Bolt12SemanticError::InvalidAmount`

Added doc update to clarify that `InvalidAmount` now also covers
cases where the amount in an invoice does not match the expected
amount specified in the associated `InvoiceRequest` or `Refund`.

slanesuke committed:
Add test for refund creation with insufficient liquidity

Adds the `fails_creating_refund_with_insufficient_liquidity` test to
verify that the `create_refund_builder` method correctly handles
cases where there is insufficient channel liquidity. The test verifies
that the method returns the `Bolt12RequestError::InsufficientLiquidity`
error when the refund amount exceeds the available liquidity.

slanesuke committed:
f Remove currency validation from parsing code

This fixup commit removes the currency validation from the
`InvoiceRequest` parsing code. Additionally, updates a test
that previously expected a failure due to an unsupported
currency. Since the logic now parses any amount the test no
longer triggers the `UnsupportedCurrency` error during parsing.

The next commit adds the currency validation during handling
in `ChannelManager` instead. This way, when currency support is
added, we avoid the need for exchange rate lookups at parsing
time.

slanesuke committed:
Move currency check to `OffersMessageHandler`

Add and move the check for unsupported currencies from
the parsing code to `handle_message` in `ChannelManager`.
This change ensures that we don't fail when parsing an
`InvoiceRequest` for an Offer with a currency.

slanesuke commented:
rebased

shaavan commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -1598,23 +1599,23 @@ fn fails_creating_or_paying_for_offer_without_connected_peers() {
-		.amount_msats(10_000_000)
-		.build().unwrap();
+		.create_offer_builder(None).unwrap()
+		.amount(Amount::Currency { iso4217_code: *b"USD", amount: 6_000 })
```
Since we are creating offers in fiat denomination in [Add tests for Offers with fiat amount](https://github.com/lightningdevkit/rust-lightning/pull/3175/commits/eee5eeb1a2b4ddc6a5198fe527a306cbeb22e22c) commit, maybe we can remove this change and the change below to simplify the PR?

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2548,6 +2546,40 @@ pub enum RecentPaymentDetails {
 	},
 }
 
+/// Error during creation and handling of BOLT 12 related payments.
```
The roles of Bolt12CreationError, and Bolt12Response feels a little confusing to me.

<details>
<summary>
For example
</summary>

It reads _during creation and handling of BOLT 12 related payments_ but technically we are not creating any payment when we are creating an offer.

</details>


Maybe we can simplify the enums, by introducing only one new enum called `Bolt12CreationError` that contains "Error during creation of BOLT12 Messages"?

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8952,7 +8984,28 @@ macro_rules! create_refund_builder { ($self: ident, $builder: ty) => {
-			.map_err(|_| Bolt12SemanticError::MissingPaths)?;
+			.map_err(|()| Bolt12RequestError::BlindedPathCreationFailed)?;
+
+		let total_liquidity: u64 = {
```
@jkczyz 

Should we add a similar liquidity check for inbound liquidity? This would be helpful when we are creating an offer, or calling `request_refund_payment`.

jkczyz replied:
Offers can be long-lived and be paid more than once. So it would be only be accurate for short-lived, single-use offers. For those and refund payment request, I wonder if that approach is preferable as it allows us to give early indication of a failure to the user. Drawback is it would prevent creating an offer / sending an invoice even if liquidity  is soon-but-not-yet available (e.g., future JIT channel work). @tnull Thoughts?

slanesuke replied:
So IIUC, the outbound liquidity check for refunds makes sense since they’re likely to be claimed quickly. But for inbound liquidity, JIT channels could complicate things. Should we hold off on adding an inbound liquidity check in this PR? I’d appreciate any thoughts on how to proceed!

jkczyz replied:
Yeah, let's hold off.

shaavan reviewed (commented):
Generally, the PR looks good to me!

Some small suggestions from my side

slanesuke commented:
rebased
